### ISSUE TITLE
Problem passing multiple spring profiles as environment variable in k8s deployer

### ISSUE THREAD

**Issue by chrisjs** [spring-cloud/spring-cloud-deployer-kubernetes#219]
Moved from https://github.com/spring-cloud/spring-cloud-dataflow-server-kubernetes/issues/276

---

We are running SCDF on Kubernetes and use a Config Server to supply configuration to stream apps (eg the rabbit location and credentials). To do this we use the following:

deployer.http.kubernetes.environmentVariables=SPRING_CLOUD_CONFIG_URI=<my_uri>,SPRING_PROFILES_ACTIVE=production

in our SCDF configuration.Recently we have had to add a second profile ie

deployer.http.kubernetes.environmentVariables=SPRING_CLOUD_CONFIG_URI=<my_uri>,SPRING_PROFILES_ACTIVE=production,us-west-2

But we get an error on deployment 'Invalid environment variable declared: us-west-2'

It looks like an issue with the string parsing ie we can't have a comma separated environment variable value. I've tried some simple quoting around the value but can't find a way to get this to work - is there a way to do this or is this something that needs fixed?

**Comment by chrisjs** [comment]
@djgeary after looking at this some more, the parsing logic seems fine. Variables containing commas need to be surrounded by `'`'s. We have a test for this in `KubernetesAppDeployerTests.deployWithEnvironmentWithCommaDelimitedValue()`

To get your use case to work in the CLI, using https://github.com/spring-cloud/spring-cloud-dataflow/issues/1826 as a guide and for more context on why its done this way, I did the following:

* Created a file `/tmp/env.props` wit the following content:

```
deployer.log.kubernetes.environmentVariables=SPRING_CLOUD_CONFIG_URI=http://config.com/path,SPRING_PROFILES_ACTIVE='production,us-west-2'
```

* In the data flow shell, I created a stream and read in that properties file:

```
dataflow:>stream create --name ticktock --definition "time | log"
Created new stream 'ticktock'
dataflow:>stream deploy ticktock --propertiesFile "/tmp/env.props"
Deployment request has been sent for stream 'ticktock'
```

* I then verified `SPRING_CLOUD_CONFIG_URI` and `SPRING_PROFILES_ACTIVE` were set by doing a `describe` on the deployed `ticktock` pod and confirmed the following:

```
    Environment:
      SPRING_CLOUD_CONFIG_URI:         http://config.com/path
      SPRING_PROFILES_ACTIVE:             production,us-west-2
      ....
      ....
```

Using the above approach seems to make your use case work as expected.

**Comment by djgeary** [comment]
@chrisjs I've tried using the shell and a deployment properties file as you have suggested, and I can confirm that that approach works. 

However we are using a config server set up that contains the location of the rabbit, postgres, mongo etc that stream and task apps use so we want SPRING_CLOUD_CONFIG_URI and SPRING_PROFILES_ACTIVE set up automatically for all stream and task apps. Having to set this in a property file for each deployed app is hard to maintain and error prone. 

As described above ideally we want to set the property deployer.http.kubernetes.environmentVariables=SPRING_CLOUD_CONFIG_URI=<my_uri>,SPRING_PROFILES_ACTIVE=production,us-west-2 either in the kubernetes config map or on the config server itself, so these properties will apply to all apps.

This works fine when we only have a single SPRING_PROFILES_ACTIVE but fails with a comma separated list (even with single quotes). So there must be some difference in the parsing between the shell mechanism and using the deployer.http.kubernetes.environmentVariables server property.

A minor related issue is that there is a difference between the shell and the SCDF UI. The shell worked OK with the deployment properties file, but the UI 'Deployment Properties' box wont let be enter the deployer.http.kubernetes.environmentVariables value (I think it doesn't like the '=' in the value). The deploy button is disabled.

**Comment by chrisjs** [comment]
Hi @djgeary - understood on your preferred method of setting the properties, but the example I approach I provided above that works is based off the supported options outlined in https://github.com/spring-cloud/spring-cloud-dataflow/issues/1826

In regards to shell parsing, you may want to followup on https://github.com/spring-cloud/spring-cloud-dataflow/issues/1826 with your use case. If your running into issues with the SCDF UI, you may want to log an issue at: https://github.com/spring-cloud/spring-cloud-dataflow/issues

**Comment by djgeary** [comment]
@chrisjs OK - I don't think its an issue here at the deployer level as the shell approach proves it works - so its more at the SCDF level ie the original issue https://github.com/spring-cloud/spring-cloud-dataflow-server-kubernetes/issues/276

**Comment by djgeary** [comment]
Actually i think maybe the issue is in this project - the description here https://docs.spring.io/spring-cloud-dataflow-server-kubernetes/docs/1.4.0.RELEASE/reference/htmlsingle/#_environment_variables seems to cover the exact case I'm trying - and it mentions that single quotes should work - but thats not what I'm seeing when I set the property on the config server.

Is it to do with the binding in https://github.com/spring-cloud/spring-cloud-deployer-kubernetes/blob/master/src/main/java/org/springframework/cloud/deployer/spi/kubernetes/KubernetesDeployerProperties.java. Will this account for the single quotes when binding to String[]? Possibly the shell does something different here? I think there was a specific fix put in it.

Basically I get 'Invalid environment variable declared: us-west-2' implying it parses the commas irrespective of the quotes. This is on the latest version.

**Comment by djgeary** [comment]
Steps to reproduce:

Set up spring cloud dataflow kubernetes server reading config from either a config server or k8s config map, I've tried both, will use config server here.

On config server add the file spring-cloud-dataflow-server-kubernetes.yml with the following configuration

```
spring:
  cloud:
    deployer:
      kubernetes:
        environmentVariables: SPRING_CLOUD_CONFIG_URI=https://myconfigserver,SPRING_PROFILES_ACTIVE=staging-us-west-2,JAVA_TOOL_OPTIONS=-Xmx1024m

management:
  security:
    enabled: false
```

Start the SCDF server. Hitting the SCDF actuator endpoints I get 

http://myscdfserver/management/env
```
"configService:https://bitbucket.org/myorg/myrepo/spring-cloud-dataflow-server-kubernetes.yml": {
  "spring.cloud.deployer.kubernetes.environmentVariables": "SPRING_CLOUD_CONFIG_URI=https://myconfigserver,SPRING_PROFILES_ACTIVE=staging-us-west-2,JAVA_TOOL_OPTIONS=-Xmx1024m"
},
```

http://myscdfserver/management/configprops
```
"spring.cloud.deployer.kubernetes-org.springframework.cloud.deployer.spi.kubernetes.KubernetesDeployerProperties": {
....
"environmentVariables": [
  "SPRING_CLOUD_CONFIG_URI=https://myconfigserver",
  "SPRING_PROFILES_ACTIVE=staging-us-west-2",
  "JAVA_TOOL_OPTIONS=-Xmx1024m"
],
```
So this is OK.

Now change the config on the config server
```
spring:
  cloud:
    deployer:
      kubernetes:
        environmentVariables: SPRING_CLOUD_CONFIG_URI=https://myconfigserver,SPRING_PROFILES_ACTIVE='staging,us-west-2,staging-us-west-2',JAVA_TOOL_OPTIONS=-Xmx1024m    
```
Restart the SCDF server. Hitting the SCDF actuator endpoints I now get 

http://myscdfserver/management/env
```
"configService:https://bitbucket.org/myorg/myrepo/spring-cloud-dataflow-server-kubernetes.yml": {
  "spring.cloud.deployer.kubernetes.environmentVariables": "SPRING_CLOUD_CONFIG_URI=https://myconfigserver,SPRING_PROFILES_ACTIVE='staging,us-west-2,staging-us-west-2',JAVA_TOOL_OPTIONS=-Xmx1024m"
},
```

http://myscdfserver/management/configprops
```
"spring.cloud.deployer.kubernetes-org.springframework.cloud.deployer.spi.kubernetes.KubernetesDeployerProperties": {
....
"environmentVariables": [
  "SPRING_CLOUD_CONFIG_URI=https://myconfigserver",
  "SPRING_PROFILES_ACTIVE='staging",
  "us-west-2",
  "staging-us-west-2'",
  "JAVA_TOOL_OPTIONS=-Xmx1024m"
],
```
So the environmentVariables have been parsed incorrectly and will result in errors if I try to deploy apps.

The test you mention does seem to cover this exact case, but its not working on a real deployement.

**Comment by djgeary** [comment]
After looking at this some more I've found a solution that works in my setup:

If instead in the config server yml I use
```
spring:
  cloud:
    deployer:
      kubernetes:
        environmentVariables: 
          - SPRING_CLOUD_CONFIG_URI=https://myconfigserver
          - SPRING_PROFILES_ACTIVE=staging,us-west-2,staging-us-west-2
          - JAVA_TOOL_OPTIONS=-Xmx1024m
```
it works (and its clearer to read)

I thought this might imply the issue is in the way the yml file is bound to KubernetesDeployerProperties which isn't explicity covered in the test, so I tried to set SPRING_CLOUD_DEPLOYER_KUBERNETES_ENVIRONMENTVARIABLES directly as a server property (in server-deployment.yaml) but this gave the same incorrect config as before.

However we now have a working solution.
